library/crypto/sha512.c

```diff
@@ -260,22 +260,28 @@ sha512_file (int fd, off_t offset, size_t count, void *resblock)
           if (sum == BLOCKSIZE)
             break;
 
-          if (n == 0) // eof
-              goto process_partial_block;
+          if (n == 0) { /* eof */
+              if (total + sum < count) {
+                  free (buffer);
+                  return -ESPIPE;
+              } else {
+                  goto process_partial_block;
+              }
+          }
         }
 
-      /* Process buffer with BLOCKSIZE bytes.  Note that
-                        BLOCKSIZE % 128 == 0
-       */
-      if (total + BLOCKSIZE < count) {
-          total += BLOCKSIZE;
+      if (total + sum < count) {
```

Comment:
tests: added crypto test of some crypto library functions: sha512_file(), sha512_file_ctx(), sha512_buffer(): https://github.com/reverbrain/eblob/commit/b6ff3b2460e387084e6603f9126b93c2fc000285